x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java

```diff
@@ -504,4 +552,91 @@ protected static Map<String, ComposableIndexTemplate> parseComposableTemplates(I
+                            requiredPipeline.getId(),
+                            getOrigin()
+                        );
+                        creationCheck.set(false);
+                    }
+                } else {
+                    logger.debug("adding ingest pipeline [{}] for [{}], because it doesn't exist", requiredPipeline.getId(), getOrigin());
+                    putIngestPipeline(requiredPipeline, creationCheck);
+                }
+            }
+        }
+    }
+
+    @Nullable
+    private static PipelineConfiguration findInstalledPipeline(ClusterState state, String pipelineId) {
+        Optional<IngestMetadata> maybeMeta = Optional.ofNullable(state.metadata().custom(IngestMetadata.TYPE));
+        return maybeMeta.map(ingestMetadata -> ingestMetadata.getPipelines().get(pipelineId)).orElse(null);
+    }
+
+    private void putIngestPipeline(final IngestPipelineConfig pipelineConfig, final AtomicBoolean creationCheck) {
+        final Executor executor = threadPool.generic();
+        executor.execute(() -> {
+            PutPipelineRequest request = new PutPipelineRequest(pipelineConfig.getId(), pipelineConfig.loadConfig(), XContentType.JSON);
+            request.masterNodeTimeout(TimeValue.timeValueMinutes(1));
+
+            executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                getOrigin(),
+                request,
+                new ActionListener<AcknowledgedResponse>() {
+                    @Override
+                    public void onResponse(AcknowledgedResponse response) {
+                        creationCheck.set(false);
+                        if (response.isAcknowledged() == false) {
+                            logger.error(
+                                "error adding ingest pipeline [{}] for [{}], request was not acknowledged",
+                                pipelineConfig.getId(),
+                                getOrigin()
+                            );
+                        } else {
+                            logger.info("adding ingest pipeline {}", pipelineConfig.getId());
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        creationCheck.set(false);
+                        onPutPipelineFailure(pipelineConfig.getId(), e);
+                    }
+                },
+                (req, listener) -> client.execute(PutPipelineAction.INSTANCE, req, listener)
+            );
+        });
+    }
+
+    /**
+     * Called when creation of an ingest pipeline fails.
+     *
+     * @param pipelineId the pipeline that failed to be created.
+     * @param e The exception that caused the failure.
+     */
+    protected void onPutPipelineFailure(String pipelineId, Exception e) {
+        logger.error(() -> format("error adding ingest pipeline template [%s] for [%s]", pipelineId, getOrigin()), e);
```

Comment:
This is tripping up forbidden APIs check in CI - You should be able to just use string concat or use the `Strings.format` utility to handle the locale for you

Reply:
Wait, nevermind, it was in the test, will put a comment there instead. 🙈